### PR TITLE
[Quant] Allow setting fixed qparams for inner LSTM ops

### PR DESCRIPTION
**Pull request by andrewor14**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88456

Summary: In both eager and FX graph mode quantization,
`torch.ao.nn.quantizable.LSTM` is used as an observed custom module,
which is responsible for inserting its own observers. By default,
the user specifies a single QConfig for the custom module (either
through QConfigMapping or by setting the "qconfig" attribute"),
and all inner ops will [inherit this
QConfig](https://github.com/pytorch/pytorch/blob/dc00bb51b8d370bf3891f0edb2c6e0c2914e329a/torch/ao/nn/quantizable/modules/rnn.py#L366-L378)
and use the same observer/fake_quantize constructors.

Today, users who wish to override this behavior must extend
`torch.ao.nn.quantizable.LSTM` and write a lot of custom code
to manually assign the QConfigs to the inner ops. This commit
alleviates this burden on the user by providing a helper function
to assign QConfigs with custom observers. An example use case of
this is providing a reference implementation for a backend kernel
that hardcodes qparams for efficiency.

Example usage:
```
import torch
from torch.ao.quantization import get_default_qconfig_mapping
from torch.ao.quantization.fx.custom_config import (
    PrepareCustomConfig,
    ConvertCustomConfig,
)

class MyModel(torch.nn.Module):
    ...

class UserLSTM(torch.ao.nn.quantizable.LSTM):
    @classmethod
    def from_float(cls, other):
        assert isinstance(other, cls._FLOAT_MODULE)
        linear_output_obs_ctr = FixedQParamsObserver.with_args(
            scale=2 ** -11, zero_point=2 ** 15, dtype=torch.qint32)
        sigmoid_obs_ctr = FixedQParamsObserver.with_args(
            scale=2 ** -16, zero_point=0, dtype=torch.qint32)
        tanh_obs_ctr = FixedQParamsObserver.with_args(
            scale=2 ** -15, zero_point=2 ** 15, dtype=torch.qint32)
        cell_state_obs_ctr = FixedQParamsObserver.with_args(
            scale=2 ** -11, zero_point=0, dtype=torch.qint32)
        hidden_state_obs_ctr = FixedQParamsObserver.with_args(
            scale=2 ** -7, zero_point=2 ** 7, dtype=torch.quint8)
        return torch.ao.quantization.utils._get_lstm_with_individually_observed_parts(
            float_lstm=other,
            linear_output_obs_ctr=linear_output_obs_ctr,
            sigmoid_obs_ctr=sigmoid_obs_ctr,
            tanh_obs_ctr=tanh_obs_ctr,
            cell_state_obs_ctr=cell_state_obs_ctr,
            hidden_state_obs_ctr=hidden_state_obs_ctr,
        )

qconfig_mapping = get_default_qconfig_mapping()
example_inputs = (torch.rand(5, 3, 50), torch.rand(1, 3, 50), torch.randn(1, 3, 50))
prepare_custom_config = PrepareCustomConfig() \
    .set_float_to_observed_mapping(torch.nn.LSTM, UserLSTM)
convert_custom_config = ConvertCustomConfig() \
    .set_observed_to_quantized_mapping(torch.ao.nn.quantizable.LSTM, torch.ao.nn.quantized.LSTM)
model = MyModel()
model = prepare_fx(model, qconfig_mapping, example_inputs, prepare_custom_config=prepare_custom_config)
model(*example_inputs)  # calibrate
model = convert_fx(model, convert_custom_config=convert_custom_config)
model(*example_inputs)
```

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_static_lstm_with_custom_fixed_qparams

Reviewers: jerryzh168, vkuzo

Subscribers: jerryzh168, vkuzo

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @Xia-Weiwen @leslie-fang-intel

### CODE REVIEW

File: torch/ao/quantization/quantize.py [hunk 1]
```diff
@@ -214,12 +214,12 @@ def insert_activation_post_process(m, special_act_post_process=None):
...
                 insert_activation_post_process(child)
-        elif _has_special_act_post_process(child):
-            special_act_post_process = _get_special_act_post_process(child)
-            insert_activation_post_process(child, special_act_post_process)
         elif non_leaf_module_list is not None and type_before_parametrizations(child) in non_leaf_module_list:
             if needs_observation(child):
                 insert_activation_post_process(child)
+        elif _has_special_act_post_process(child):
+            special_act_post_process = _get_special_act_post_process(child)
+            insert_activation_post_process(child, special_act_post_process)
```
Comment (andrewor14) [inline torch/ao/quantization/quantize.py:222]: I plan to separate these changes to a different PR to better explain why this is needed

**Comment by vkuzo** [comment]
What's the current default on how the user is expected to insert observers in a custom module?  Can we link to the docs on how that is supposed to work before this PR?

**Comment by andrewor14** [comment]
> What's the current default on how the user is expected to insert observers in a custom module? Can we link to the docs on how that is supposed to work before this PR?

Added a link in the PR description

File: torch/ao/quantization/quantize.py (hunk 1 above)
Comment (andrewor14) [reply torch/ao/quantization/quantize.py:222]: Moved these changes to here: https://github.com/pytorch/pytorch/pull/88498. Please review that one first

File: torch/ao/quantization/utils.py [hunk 2]
```diff
@@ -547,6 +547,89 @@ def _patched_module_call(self, *args, **kwargs):
     return fqn_to_example_inputs
 
 
+def _get_fixed_params_observed_lstm(
+    float_lstm: torch.nn.LSTM,
+    is_qat: bool,
+    linear_output_qparams: Optional[Tuple[float, int]] = None,
```
Comment (vkuzo) [inline torch/ao/quantization/utils.py:553]: can we ask the user to pass in an observer/fake_quant constructor instead, so it works the same way as other APIs?

```
obs_ctr = FakeQuantize.with_args(...)
_get_lstm_with_individually_observed_parts(
    float_lstm,
    linear_obs_ctr,
    ...
)
```

File: torch/ao/quantization/utils.py [hunk 3]
```diff
@@ -547,6 +547,89 @@ def _patched_module_call(self, *args, **kwargs):
     return fqn_to_example_inputs
 
 
+def _get_fixed_params_observed_lstm(
```
Comment (vkuzo) [inline torch/ao/quantization/utils.py:550]: can we provide a module the user can use directly instead of providing a util function?  Or, can we just make the quantizeable module do this by default?

Comment (vkuzo) [reply torch/ao/quantization/utils.py:550]: behavior which would make sense to me from user's perspective:
* user is quantizing for backend X (default: fbgemm/qnnpack)
* user is using a get_default_qconfig_mapping(backend_str) function
* for static quant, the right qconfig needs to be passed to LSTM
* if LSTM is using custom modules, we need to follow the custom module spec in how things are assigned qconfigs and/or quantized.  Does it even make sense at all to "assign parent qconfig to all child modules" for custom modules?

we should really understand the custom module spec and role in quantization (where is that defined)?

option 1:
* LSTM keeps using custom modules, we assign qconfig according to custom module spec (parent framework never assigns qconfig of insides of custom module, that is the user's job)

option 2:
* LSTM does not use custom modules, and then it can use qconfig_mapping

Comment (jerryzh168) [reply torch/ao/quantization/utils.py:550]: option 1 right now I think, we are delaying the design of custom module in reference quantized model flow due to PT2 export, I think we can come up with the design in PT2 export

**Review by jerryzh168 — APPROVED** [review]
LGTM

File: torch/ao/quantization/utils.py (hunk 3 above)
Comment (andrewor14) [reply torch/ao/quantization/utils.py]: As discussed offline, the problem with the current LSTM flow (before this PR) is that the user can specify the QConfig for `torch.nn.LSTM` through QConfigMapping, but this breaks the custom module abstraction where the custom module itself is responsible for inserting its own observers. The new flow enabled in this PR respects this abstraction.

To fix the default behavior, we can check QConfigMapping and throw an exception if we find `torch.nn.LSTM` (and MHA) with a helpful message to redirect users towards the right behavior, which is to manually set (`model.lstm.qconfig = QConfig(...)`), which respects the custom module abstractions. Since this is technically a separate issue, it will be fixed in a future PR.

**Comment by andrewor14** [comment]
Ok, I think I have addressed all the comments. Fixing the existing default behavior with `torch.ao.nn.quantizable.LSTM` to make it respect the custom module abstractions will be handled in a future PR. This PR only adds an alternative to the existing behavior.

**Review by vkuzo — APPROVED** [review]
